### PR TITLE
chore(tests): Change integration test suite to use poetry

### LINKED ISSUE #10907: Update integration tests to use poetry
┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-4644)

### PR DESCRIPTION
**Pull request by jrbenny35**
Because

- We use poetry for most other package management

This commit

- Updates our integration tests to use poetry too.

Fixes #10907

### CODE REVIEW

File: experimenter/tests/integration/legacy/test_filter_expression_validation.py [hunk 1]
```diff
@@ -1,14 +1,29 @@
-def test_filter_expressions_with_matching_firefox_versions(base_url, selenium):
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture
+def filter_expression_path():
+    here = Path(__file__).cwd()
```
Comment (brennie) [inline experimenter/tests/integration/legacy/test_filter_expression_validation.py:8]: This is wrong.

`Path.cwd()` returns the current working directory. `Path(xyz).cwd()` will first create a path pointing to `xyz` and then throw it out and return the current working directory.

If you want the parent directory of `__file__`, its `Path(__file__).parent()`.

File: experimenter/tests/integration/legacy/test_filter_expression_validation.py [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+    return Path(next(iter(here.glob("**/legacy/utils/filter_expression.js"))))
```
Comment (brennie) [inline experimenter/tests/integration/legacy/test_filter_expression_validation.py:9]: We should just hardcode the full path so we dont have to glob.

File: experimenter/tests/integration/legacy/test_filter_expression_validation.py [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
+
+
+def test_filter_expressions_with_matching_firefox_versions(
+    base_url, selenium, filter_expression_path
+):
     selenium.get("about:blank")
-    with open("legacy/utils/filter_expression.js") as js:
+    with Path(filter_expression_path).open() as js:
```
Comment (brennie) [inline experimenter/tests/integration/legacy/test_filter_expression_validation.py:16]: `filter_expression_path` is already a `Path`, so you can call `.open()` directly.

File: experimenter/tests/integration/legacy/test_filter_expression_validation.py [hunk 4]
```diff
@@ -1,14 +1,29 @@
...
     assert script is True
 
 
-def test_filter_expressions_with_mismatching_firefox_versions(base_url, selenium):
+def test_filter_expressions_with_mismatching_firefox_versions(
+    base_url, selenium, filter_expression_path
+):
     selenium.get("about:blank")
-    with open("legacy/utils/filter_expression.js") as js:
+    with Path(filter_expression_path).open() as js:
```
Comment (brennie) [inline experimenter/tests/integration/legacy/test_filter_expression_validation.py:26]: `filter_expression_path` is already a `Path`, so you can call `.open()` directly.

File: experimenter/tests/integration/nimbus/conftest.py [hunk 5]
```diff
@@ -180,9 +180,15 @@ def experiment_url(base_url, experiment_slug):
...
     base_path = Path(
-        f"{current_path.parent.parent}/experimenter/outcomes/metric-hub-main/jetstream/outcomes"
+        next(
+            iter(
+                current_path.glob(
+                    "**/experimenter/outcomes/metric-hub-main/jetstream/outcomes"
+                )
+            )
+        )
     )
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/conftest.py:192]: Lets hardcode the path. This is more inscrutible than it has to be.

File: experimenter/tests/integration/nimbus/test_desktop_targeting.py [hunk 6]
```diff
@@ -13,12 +13,19 @@ def targeting_config_slug(request):
     return request.param
 
 
+@pytest.fixture
+def filter_expression_path():
+    here = Path(__file__).cwd()
+    return Path(next(iter(here.glob("**/nimbus/utils/filter_expression.js"))))
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_desktop_targeting.py:19]: this too.

File: experimenter/tests/integration/nimbus/test_desktop_targeting.py [hunk 7]
```diff
@@ -13,12 +13,19 @@ def targeting_config_slug(request):
     return request.param
 
 
+@pytest.fixture
+def filter_expression_path():
+    here = Path(__file__).cwd()
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_desktop_targeting.py:18]: see other comment about cwd vs parent

File: experimenter/tests/integration/nimbus/test_desktop_targeting.py [hunk 8]
```diff
@@ -32,7 +39,7 @@ def test_check_advanced_targeting(
 
     # Inject filter expression
     selenium.get("about:blank")
-    with Path("nimbus", "utils", "filter_expression.js").open() as js:
+    with Path(filter_expression_path).open() as js:
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_desktop_targeting.py:42]: `filter_expression_path` is already a `Path`, so you can call `.open()` directly.

File: experimenter/tests/integration/nimbus/test_desktop_targeting.py [hunk 9]
```diff
@@ -84,7 +92,7 @@ def test_check_audience_targeting(
...
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_desktop_targeting.py:95]: `filter_expression_path` is already a `Path`, so you can call `.open()` directly.

File: experimenter/tests/integration/nimbus/test_experimenter_ui.py [hunk 10]
```diff
@@ -16,6 +16,12 @@
 ]
 
 
+@pytest.fixture
+def image_path():
+    here = Path(__file__).cwd()
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_experimenter_ui.py:21]: See other comment about cwd vs parent

File: experimenter/tests/integration/nimbus/test_experimenter_ui.py [hunk 11]
```diff
@@ -16,6 +16,12 @@
...
+    return Path(next(iter(here.glob("**/example.jpg"))))
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_experimenter_ui.py:22]: Lets hardcode this too

File: experimenter/tests/integration/nimbus/test_mobile_targeting.py [hunk 12]
```diff
@@ -29,7 +29,8 @@ def record_malformed_feature_config(self, *args, **kwargs):
 
 
 def client_info_list():
-    with Path("nimbus", "app_contexts.json").open() as file:
+    here = Path(__file__).cwd()
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_mobile_targeting.py:32]: see other commend about cwd vs parent

File: experimenter/tests/integration/nimbus/test_mobile_targeting.py [hunk 13]
```diff
@@ -29,7 +29,8 @@ def record_malformed_feature_config(self, *args, **kwargs):
...
+    with Path(next(iter(here.glob("**/nimbus/app_contexts.json")))).open() as file:
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/test_mobile_targeting.py:33]: see other comments about hardcoding paths

File: experimenter/tests/nimbus_integration_tests.sh [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+set +x
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+if [[ -n "${UPDATE_FIREFOX_VERSION}" ]]; then
+    sudo ./experimenter/tests/integration/nimbus/utils/nightly-install.sh
+fi
+
+sudo apt-get -qqy update && sudo apt-get -qqy install python3-venv python3-pip
+pip install poetry --break-system-packages
```
Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:13]: Neat, i've neve seen `--break-system-packages` before. I think that maybe warrants a comment

File: experimenter/tests/nimbus_integration_tests.sh [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+sudo chmod -R a+rwx /code/experimenter/tests/integration/
+mkdir -m a+rwx /code/experimenter/tests/integration/test-reports
+
+poetry -C experimenter/tests/integration -vvv install --no-root
+poetry -C experimenter/tests/integration -vvv run pytest --html=experimenter/tests/integration/test-reports/report.htm --self-contained-html --reruns-delay 30 --driver Firefox --verify-base-url --base-url https://nginx/nimbus/ experimenter/tests/integration/nimbus -vvv $PYTEST_ARGS
```
Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:18]: Can we break this into multiple lines?
```suggestion
poetry -C experimenter/tests/integration
       -vvv \
       run \
       pytest \
       --html=experimenter/tests/integration/test-reports/report.htm \
       --self-contained-html \
       --reruns-delay 30 \
       --driver Firefox \
       --verify-base-url \
       --base-url https://nginx/nimbus/ \
       experimenter/tests/integration/nimbus \
       -vvv \
       $PYTEST_ARGS
```

Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:18]: -vvv is in here twice?

File: experimenter/tests/nimbus_rust_tests.sh [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+set +x
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+apt-get -qqy update && apt-get -qqy install python3-venv python3-pip
+pip install poetry
+mkdir -m a+rwx /code/experimenter/tests/integration/test-reports
+
+poetry -C experimenter/tests/integration -vvv install --no-root
+poetry -C experimenter/tests/integration run pytest --verify-base-url --base-url https://nginx/nimbus/ --html=test-reports/report.htm --self-contained-html experimenter/tests/integration/nimbus/test_mobile_targeting.py -vvv
```
Comment (brennie) [inline experimenter/tests/nimbus_rust_tests.sh:13]: Can we break this up too?

File: experimenter/tests/nimbus_integration_tests.sh (hunk 14 above)
Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:13]: This should maybe live in the dockerfile?

File: experimenter/tests/integration/nimbus/conftest.py [hunk 17]
```diff
@@ -180,9 +180,15 @@ def experiment_url(base_url, experiment_slug):
 def fixture_load_experiment_outcomes():
     """Fixture to create a list of outcomes based on the current configs."""
     outcomes = {"firefox_desktop": "", "fenix": "", "firefox_ios": ""}
-    current_path = Path.cwd()
+    current_path = Path().cwd()
```
Comment (brennie) [inline experimenter/tests/integration/nimbus/conftest.py:183]: Path.cwd() was correct. its a classmethod

File: experimenter/tests/experimenter_legacy_tests.sh [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+sudo apt-get -qqy update && sudo apt-get -qqy install python3-venv python3-pip
+pip install poetry --break-system-packages
+sudo chmod -R a+rwx /code/experimenter/tests/integration/
+mkdir -m a+rwx /code/experimenter/tests/integration/test-reports
+
+poetry -C experimenter/tests/integration -vvv install --no-root
+poetry -C experimenter/tests/integration -vvv run pytest --html=experimenter/tests/integration/test-reports/report.htm --self-contained-html --reruns-delay 30 --driver Firefox --verify-base-url --base-url https://nginx/nimbus/ experimenter/tests/integration/legacy/ -vvv
```
Comment (brennie) [inline experimenter/tests/experimenter_legacy_tests.sh:14]: break this across multiple lines too

**Review by brennie — CHANGES_REQUESTED** [review]

File: experimenter/tests/nimbus_integration_tests.sh (hunk 15 above)
Comment (jrbenny35) [reply experimenter/tests/nimbus_integration_tests.sh:18]: One if for poetry the other for the pytest command

File: experimenter/tests/nimbus_integration_tests.sh (hunk 14 above)
Comment (jrbenny35) [reply experimenter/tests/nimbus_integration_tests.sh:13]: Good point!

**Comment by jrbenny35** [comment]
Thanks @brennie, Pretty obvious to see I'm still new at using `pathlib` but this has been great, learned a lot.

File: experimenter/tests/experimenter_legacy_tests.sh [hunk 19]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -euo pipefail
+set +x
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+sudo apt-get -qqy update && sudo apt-get -qqy install python3-venv python3-pip
+pip install poetry --break-system-packages # https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-break-system-packages
```
Comment (brennie) [inline experimenter/tests/experimenter_legacy_tests.sh:9]: Sorry, I meant a comment about *why* we're doing --break-system-packages

File: experimenter/tests/experimenter_legacy_tests.sh [hunk 20]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -euo pipefail
+set +x
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+sudo apt-get -qqy update && sudo apt-get -qqy install python3-venv python3-pip
```
Comment (brennie) [inline experimenter/tests/experimenter_legacy_tests.sh:8]: Isnt this in the dockerfile now?

File: experimenter/tests/nimbus_integration_tests.sh [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+set -euo pipefail
+set +x
+
+export PATH=$PATH:/home/seluser/.local/bin
+
+if [[ -n "${UPDATE_FIREFOX_VERSION}" ]]; then
+    sudo ./experimenter/tests/integration/nimbus/utils/nightly-install.sh
+fi
+
+sudo apt-get -qqy update && sudo apt-get -qqy install python3-venv python3-pip
```
Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:12]: this is in the dockerfile isnt it?

File: experimenter/tests/nimbus_integration_tests.sh [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+pip install poetry --break-system-packages # https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-break-system-packages
```
Comment (brennie) [inline experimenter/tests/nimbus_integration_tests.sh:13]: see other comment on other script

**Review by brennie — APPROVED** [review]

File: experimenter/tests/experimenter_legacy_tests.sh (hunk 20 above)
Comment (jrbenny35) [reply experimenter/tests/experimenter_legacy_tests.sh:8]: No, we don't control this dockerfile, its built directly from selenium source

**Comment by jaredlockhart** [comment]
Heyyyy!!!!  Look at this!! Now we can bump from 3.8 to 3.11??? Amazing thank you @jrbenny35 🎉 🎉 🎉 🎉 🎉 🎉

**Pull request merged**